ts/async_service.py

```diff
@@ -0,0 +1,156 @@
+from threading import Thread
+
+from ts.handler_utils.utils import create_predict_response
+from ts.protocol.otf_message_handler import retrieve_msg
+from ts.service import PREDICTION_METRIC, PredictionException, Service
+
+logger = logging.getLogger(__name__)
+
+
+async def predict(self, batch):
+    """
+    PREDICT COMMAND = {
+        "command": "predict",
+        "batch": [ REQUEST_INPUT ]
+    }
+    :param batch: list of request
+    :return:
+
+    """
+    headers, input_batch, req_id_map = Service.retrieve_data_for_inference(batch)
+
+    context = copy.deepcopy(self.context)
+
+    context.request_ids = req_id_map
+    context.request_processor = headers
+    context.cl_socket = self.cl_socket
+    metrics = context.metrics
+    metrics.request_ids = req_id_map
+
+    start_time = time.time()
+
+    # noinspection PyBroadException
+    try:
+        print(f"{self._entry_point=}")
+        ret = await self._entry_point(input_batch, context)
+    except MemoryError:
+        logger.error("System out of memory", exc_info=True)
+        return create_predict_response(None, req_id_map, "Out of resources", 507)
+    except PredictionException as e:
+        logger.error("Prediction error", exc_info=True)
+        return create_predict_response(None, req_id_map, e.message, e.error_code)
+    except Exception as ex:  # pylint: disable=broad-except
+        if "CUDA" in str(ex):
+            # Handles Case A: CUDA error: CUBLAS_STATUS_NOT_INITIALIZED (Close to OOM) &
+            # Case B: CUDA out of memory (OOM)
+            logger.error("CUDA out of memory", exc_info=True)
+            return create_predict_response(None, req_id_map, "Out of resources", 507)
+        else:
+            logger.warning("Invoking custom service failed.", exc_info=True)
+            return create_predict_response(None, req_id_map, "Prediction failed", 503)
+
+    if not isinstance(ret, list):
+        logger.warning(
+            "model: %s, Invalid return type: %s.",
+            context.model_name,
+            type(ret),
+        )
+        return create_predict_response(
+            None, req_id_map, "Invalid model predict output", 503
+        )
+
+    if len(ret) != len(input_batch):
+        logger.warning(
+            "model: %s, number of batch response mismatched, expect: %d, got: %d.",
+            context.model_name,
+            len(input_batch),
+            len(ret),
+        )
+        return create_predict_response(
+            None, req_id_map, "number of batch response mismatched", 503
+        )
+
+    duration = round((time.time() - start_time) * 1000, 2)
+    metrics.add_time(PREDICTION_METRIC, duration)
+
+    return create_predict_response(
+        ret, req_id_map, "Prediction success", 200, context=context
+    )
+
+
+class AsyncService(object):
+    def __init__(self, service):
+        self.service = service
+        self.service.predict = types.MethodType(predict, self.service)
+        self.in_queue = Queue()
```

Comment:
Curious why we need a thread safe queue for the `in_queue` or we could safely use the non thread-safe async queue, similar to `out_queue` here as well?

Reply:
Yes, thats a good point. We need the thread safe queue here as async queue does not provide a timeout. The timeout is actually not strictly necessary for the vllm use case but I wanted to recreate the same batching functionality as in the frontend for future reference. If e.g. someone wants to run multiple workers with batching in the backend.